src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,83 +1080,81 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
 void MetaBlkService::recover(bool do_comp_cb) {
     // for each registered subsystem, look up in cache for their meta blks;
     std::lock_guard< decltype(m_shutdown_mtx) > lg{m_shutdown_mtx};
-    for (auto& m : m_meta_blks) {
-        auto* mblk = m.second;
-        auto buf = read_sub_sb_internal(mblk);
-
-        // found a meta blk and callback to sub system;
-        const auto itr = m_sub_info.find(mblk->hdr.h.type);
-        if (itr != std::end(m_sub_info)) {
-            // if subsystem registered crc protection, verify crc before sending to subsystem;
-            if (itr->second.do_crc) {
-                const auto crc = crc32_ieee(init_crc32, s_cast< const uint8_t* >(buf->bytes), mblk->hdr.h.context_sz);
-
-                HS_REL_ASSERT_EQ(crc, uint32_cast(mblk->hdr.h.crc),
-                                 "[type={}], CRC mismatch: {}/{}, on mblk bid: {}, context_sz: {}", mblk->hdr.h.type,
-                                 crc, uint32_cast(mblk->hdr.h.crc), mblk->hdr.h.bid.to_string(),
-                                 uint64_cast(mblk->hdr.h.context_sz));
-            } else {
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk found with bypassing crc.", mblk->hdr.h.type);
-            }
+    for (const auto& subtype : m_subtype_recovery_order) {
```

Comment:
Agree to recovery by subtype however doesnt understand why you think 1 is better than 2
1 . subtype order based on appear order in the metablk chain
2. subtype order based on the order of subtype_name(string).

As 2 by nature can be done by something like
```
for (const auto& sub : m_sub_info)
    for (const auto& sub : sub->second().meta_bids)
```

Reply:
And also not sure how this could help HO , i.e how can we achieve
```
we actually want to recover repl-dev before PGinfo and recover PGinfo before Shardinfo.
```
It seems to me neither  `metablk chain` order nor `subtype dictionary order` is the order we want (dependency order),  unless we explicitly pass-in recovery-order(from hard-code or config).

Reply:
>why you think 1 is better than 2

suppose we have two subsystems in metaservice , the name is `A` and `B`, respectively.
if we add a metablk B1, then add a metablk A1, then A2, then B2. finally,  the metablk chain will look like
`ssb->B1->A1->A2->B2`

the first metablk of subsystem B (B1) comes before the first metablk of subsystem A (A1), so we want to recover all the metablks of B before A

`client_info_map_t m_sub_info ` is a `std::map` 
```
std::map< meta_sub_type, MetaSubRegInfo > client_info_map_t;
```
and, all the items in it will be ordered by the key(A<B). so if we use `for (const auto& sub : m_sub_info)` as you mentioned, subsystem A will be recovered before subsystem B, that is not what we want here.

>unless we explicitly pass-in recovery-order

as i mentioned above, actually the recovery order of all the subsystem here will be the order of the presence of the first metablk of each subsystem in the metablk chain. so if shardInfo depends on PGinfo, PGinfo will be created before shardInfo and thus the metablk of PGinfo will be previous of shardinfo, and then the subsystem of PG will be recovered before shard.

pass-in recovery-order in unnecessary here. if we do this , the user of homestore will have to take the responsibility to pass in the correct order, which is not what we want. So , unless we want to recovery shardInfo before PGInfo, for example, we need a mechanism to specify the particular order of subsystem recovery. but i do not think there will be a scenario like this